### PR TITLE
Opmrun 2019-04

### PR DESCRIPTION
**Pull request by OPMUSER**
Delete branch 2019-04-1 and created 2019-04 for all changes and update source.

### CODE REVIEW

File: opmrun/README.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+Written in Python 3 and tested under Unbuntu-Mate 18.04 TLS.
+
+Compiled binary version should work on all Linux systems, no need to install dependencies or Python.
+
+Notes:
+------
+PySimpleGUI is the GUI tool used to build OPMRUN. It is in active development and is frequently updated for fixes and new features. This version of OPMRUN used verion 3.36.0 of PySimpleGUI, later and older   versions of PySimpleGUI may not work. Each release of OPMRUN will update to the latest release of PySimpleGUI. In addition, the following modules are require:
+            
+( 1) PySimpleGUI
+( 2) datetime
```
Comment (joakim-hove) [inline opmrun/README.md]: Observe that the modules `datetime, os, sys, getpass, re` and `subprocess` are part of the standard library - no reason to list them as dependencies.

**Comment by joakim-hove** [comment]
David; this is your project - and you decide - but *iff* you want others to make changes, and use the `opmrun` tool I would reccomend some changes:

1. As already indicated - I would avoid using `pathlib` - that way you will work even with Python2.7.
2. Again - I would avoid using the Pyinstaller.
3. Documentation should be in a plain text format - either just text or some markdown (e.g. .rst) embedded pdf is not developer friendly.
4. Embedding a large base64 enocoded string in the main program makes it very uncomfortable to edit - at least in emacs.

But again - this is your project, maybe you are not at all interested in feedback, and that is of course 100% legitimate - then I will just shut up.

**Comment by OPMUSER** [comment]
> I have looked a bit more through the opmrun script today. I have two suggestions:
> 
> As I already mentioned I really think you should use the normal setup.py based installation, instead of pyinstaller - for one thing you avoid the need to embed the binary in the source code, but more importantly you are more certain that things will just work. My experience with pyinstaller and similar tools is that they are a bit fragile. setup.py is not a very comfortable tool, but you have a quite simple - pure Python project and then setup.py does the job.
> 
> You have made quite extensive use of the Path class from the pathlib module; the pathlib module came in python version 3.4. I think the use of Path/pathlib is the most serious obstacle to Python2.7 compatability (which is unfortunately still quite relevant). If I were you I would avoid using Path/pathlib - if you are interested I can implement a small Path class for you - which will give you everything except the / operator - that must be performed with os.path.join( ... ).
> 

@joakim-hove very much appreciated your feedback and guidance.

I'll investigate creating a standard install using setup.py as part of the code base. However, I would still also like to include the binary as well, as this application is meant for end users, similarly what the project does for OPM Flow, that is we don't expect engineers to build OPM Flow from source. So I'm attempting to do the same.

I'll investigate options for Python 2 support including the Path class code you generously provided.

File: opmrun/README.md (hunk 1 above)
Comment (OPMUSER) [reply opmrun/README.md]: I'll update the documentation accordingly.

**Comment by OPMUSER** [comment]
> David; this is your project - and you decide - but if you want others to make changes, and use the opmrun tool I would recommend some changes:  
> 
> 1. As already indicated - I would avoid using pathlib - that way you will work even with Python2.7. 
> 2. Again - I would avoid using the Pyinstaller. 
> 3. Documentation should be in a plain text format - either just text or  some markdown (e.g. .rst) embedded pdf is not developer friendly. 
> 4. Embedding a large base64 enocoded string in the main program makes it very uncomfortable to edit - at least in emacs.
> 
> But again - this is your project, maybe you are not at all interested  in feedback, and that is of course 100% legitimate - then I will just  shut up/
> 
First of all I would  like to express my appreciation for the time you have taken to review and to offer guidance, it is very much appreciated.

Concerning the Python 2 issue, I have elected to use pathlib2 module that is a back port for pathlib, this should keep all the code the same and should work for both versions of Python. I have a small if statement to test for the Python version the code is running under so that the correct module is imported. I think this is a more robust solution than adding the Path class that you suggested, as path manipulation is prevalent throughout the code.

I agree with adding  setup.py option and will get to this when I more fully understand how this works. I would also like to keep the Pyinstaller version as a simple standalone solution for the engineers, similar to what we do for OPM Flow. Note what I would like to be able to do in the future is for OPMRUN to install under Windows, and to run OPM Flow through QEMU. I don't know if this can be done, but that is the long term plan. At the moment I'm concentrating on building GUI for OPM Flow on Linux, so that the engineers have similar functionality as the commercial software.

I'll also update the documentation so it is "developer friendly" but will also keep the "user friendly" pdf file.

Yes, I know embedding a large base64 enocoded string in the main program makes it very uncomfortable to edit, so I'm open to suggestions. This is what the developer for PySimpleGUI did, as it keeps everything in the one file, so I just followed his example. If there is another way to do this and keep everything together, then I'll seriously consider it.

And lastly, again, I would  like to express my appreciation for the time you have taken to review and to offer guidance, it is very much appreciated.

**Comment by akva2** [comment]
If you create a setup.py system, it can be installed with pip and we can push it to repos, put it in rpms, debs etc. You can also have the pyinstaller thing but binary should not be hosted in git. We can host it as a release on github and link from opm-project.org etc but we do not want it in the source repo.

**Comment by OPMUSER** [comment]
@akva2 that sounds like a plan, thank you.

**Comment by joakim-hove** [comment]
>I have elected to use pathlib2 module that is a back port for pathlib, this should keep all the code the same and should work for both versions of Python.

That sound like a good plan; instead of checking Python version explicitly the common way to do this is:
```
try:
     from pathlib import Path
except ImportError:
     from pathlib2 import Path
```
I reccomend the same for the GUI library.

**Comment by joakim-hove** [comment]
> [...] so I'm open to suggestions.

I would store the png image in a separate file alongside with the script[1]. The path to the current running script can be found with:
```
path = os.path.dirname( __file__ )
```


[1]: If you go the `setup.py` route - and also split the script in a package with modules and a short script, this will become more natural.

**Comment by OPMUSER** [comment]
> That sound like a good plan; instead of checking Python version explicitly the common way to do this is:
> 
> try:
>      from pathlib import Path
> except ImportError:
>      from pathlib2 import Path
> I recommend the same for the GUI library.

I kept the version check as is but added the try to check that the import works for a given version of Python. If it fails to import exit with a message - I think this more robust combining the two methods.

**Comment by OPMUSER** [comment]
> I would store the png image in a separate file alongside with the script[1]. The path to the current running script can be found with:
> 
> path = os.path.dirname( __file__ )
> image_file = os.path.join( path, "image.png")

I tried this but it does not work. This may be a bug with PySimpleGUI so I have reported the issue as a potential bug. I will **HOLD** for the moment to see if this is indeed a bug.

The current version is still okay, no binary, it is just missing the last two modifications.

**Comment by OPMUSER** [comment]
I can confirm that using the "image.png" for the icon on a window in PYSimpleGUI does not work under Linux, but does for Windows. We can either use the current base64 solution or wait for a fix. Waiting for a fix means using the latest release of PySimpleGUI, which will have to be tested before I can issue a new release. This may take some time.

**Comment by OPMUSER** [comment]
Any reason why Opmrun 2019-04 #43 cannot be merged?

**Pull request closed without merging**